*** ../corpora/repairnator-corpus/data/24/PRInformationHelperTest.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/413/PRInformationHelperTest.java	2019-10-03 10:36:38.000000000 -0400
***************
*** 77,83 ****
          baseCommit.setCommitterName("Дмитрий");
          baseCommit.setCommitterEmail("dm1998@list.ru");
          baseCommit.setMessage("Typo: fixes a typo in doc/gradle.md");
!         baseCommit.setCommittedAt(TestUtils.getDate(2016, 12, 27,14,46,43));
          baseCommit.setCompareUrl("https://github.com/INRIA/spoon/commit/2b13f0f7e82f805b2480f6ff759b2d2e8debb493");
          prInfo.setBase(baseCommit);
  
--- 77,84 ----
          baseCommit.setCommitterName("Дмитрий");
          baseCommit.setCommitterEmail("dm1998@list.ru");
          baseCommit.setMessage("Typo: fixes a typo in doc/gradle.md");
!         baseCommit.setCommittedAt(TestUtils.getDate(
!             2016, 12, 27,14,46,43));
          baseCommit.setCompareUrl("https://github.com/INRIA/spoon/commit/2b13f0f7e82f805b2480f6ff759b2d2e8debb493");
          prInfo.setBase(baseCommit);
  
