--- ../corpora/repairnator-corpus/data/24/PRInformationHelperTest.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/413/PRInformationHelperTest.java	2019-10-03 10:56:27.000000000 -0400
@@ -77,7 +77,8 @@
         baseCommit.setCommitterName("Дмитрий");
         baseCommit.setCommitterEmail("dm1998@list.ru");
         baseCommit.setMessage("Typo: fixes a typo in doc/gradle.md");
-        baseCommit.setCommittedAt(TestUtils.getDate(2016, 12, 27,14,46,43));
+        baseCommit.setCommittedAt(TestUtils.getDate(
+            2016, 12, 27,14,46,43));
         baseCommit.setCompareUrl("https://github.com/INRIA/spoon/commit/2b13f0f7e82f805b2480f6ff759b2d2e8debb493");
         prInfo.setBase(baseCommit);
 
