*** ../corpora/repairnator-corpus/data/0/LogPojo.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/74/LogPojo.java	2019-10-03 10:36:31.000000000 -0400
***************
*** 53,59 ****
          if (id != logPojo.id) return false;
          if (jobId != logPojo.jobId) return false;
          if (type != null ? !type.equals(logPojo.type) : logPojo.type != null) return false;
!         return body != null ? body.equals(logPojo.body) : logPojo.body == null;
      }
  
      @Override
--- 53,60 ----
          if (id != logPojo.id) return false;
          if (jobId != logPojo.jobId) return false;
          if (type != null ? !type.equals(logPojo.type) : logPojo.type != null) return false;
!         return body != null ? body.equals(logPojo.body) : logPojo
!                 .body == null;
      }
  
      @Override
