--- ../corpora/repairnator-corpus/data/57/BuildToBeInspected.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/389/BuildToBeInspected.java	2019-10-03 10:56:25.000000000 -0400
@@ -12,7 +12,8 @@
     // For RepairNator, the status is about the build, and for Bears it is about the previous build
     private ScannedBuildStatus status;
 
-    BuildToBeInspected(Build build, ScannedBuildStatus status) {
+    BuildToBeInspected(Build build, ScannedBuildStatus
+    status) {
         this.build = build;
         this.previousBuild = null;
         this.status = status;
