*** ../corpora/repairnator-corpus/data/20/LogParserTest.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/155/LogParserTest.java	2019-10-03 10:36:33.000000000 -0400
***************
*** 60,66 ****
          String fileContent = TestUtils.readFile(path);
          LogParser parser = new LogParser(fileContent);
  
!         assertEquals(BuildTool.GRADLE, parser.getBuildTool());
      }
  
      @Test
--- 60,67 ----
          String fileContent = TestUtils.readFile(path);
          LogParser parser = new LogParser(fileContent);
  
!         assertEquals(BuildTool.GRADLE, parser.
!         getBuildTool());
      }
  
      @Test
