--- ../corpora/repairnator-corpus/data/59/ComputeClasspath.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/146/ComputeClasspath.java	2019-10-03 10:56:12.000000000 -0400
@@ -36,7 +36,8 @@
             } catch (MalformedURLException e) {
                 this.addStepError(e.getMessage());
                 this.getLogger().error("Error while putting the following file in classpath: " + file.getAbsolutePath()
-                        + " : " + e.getMessage(), e);
+                        + " : " + e.getMessage(), e
+                    );
             }
         } else {
             this.addStepError("The file does not exist: " + file.getAbsolutePath());
