--- ../corpora/repairnator-corpus/data/31/GoogleSpreadSheetInspectorTimeSerializer.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/340/GoogleSpreadSheetInspectorTimeSerializer.java	2019-10-03 10:56:23.000000000 -0400
@@ -49,7 +49,8 @@
             int push = durations.getOrDefault(PushIncriminatedBuild.class.getSimpleName(), 0);
             int repair = durations.getOrDefault(NopolRepair.class.getSimpleName(), 0);
 
-            total = clonage + buildtime + test + gatherTestInfo + push + repair;
+            total =
+                clonage + buildtime + test + gatherTestInfo + push + repair;
             Build build = inspector.getBuild();
 
             List<Object> dataCol = new ArrayList<Object>();
