--- ../corpora/repairnator-corpus/data/75/TestAbstractStep.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/250/TestAbstractStep.java	2019-10-03 10:56:17.000000000 -0400
@@ -8,7 +8,8 @@
 
 import static org.hamcrest.CoreMatchers.is;
 import static org.junit.Assert.assertThat;
-import static org.mockito.Mockito.mock;
+import static org
+    .mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
