*** ../corpora/repairnator-corpus/data/61/TestProject.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/335/TestProject.java	2019-10-03 10:36:37.000000000 -0400
***************
*** 38,44 ****
          } else {
              if (outputTestFilter.isFailingWithTest()) {
                  this.getLogger().debug(outputTestFilter.getFailingTests() + " tests failed, go to next step.");
!                 this.state = ProjectState.TESTABLE;
              } else {
                  this.addStepError("Error while testing the project.");
                  this.shouldStop = true;
--- 38,45 ----
          } else {
              if (outputTestFilter.isFailingWithTest()) {
                  this.getLogger().debug(outputTestFilter.getFailingTests() + " tests failed, go to next step.");
!                 this.state =
!                     ProjectState.TESTABLE;
              } else {
                  this.addStepError("Error while testing the project.");
                  this.shouldStop = true;
