--- ../corpora/repairnator-corpus/data/61/TestProject.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/335/TestProject.java	2019-10-03 10:56:22.000000000 -0400
@@ -38,7 +38,8 @@
         } else {
             if (outputTestFilter.isFailingWithTest()) {
                 this.getLogger().debug(outputTestFilter.getFailingTests() + " tests failed, go to next step.");
-                this.state = ProjectState.TESTABLE;
+                this.state =
+                    ProjectState.TESTABLE;
             } else {
                 this.addStepError("Error while testing the project.");
                 this.shouldStop = true;
