--- ../corpora/repairnator-corpus/data/25/LogHelperTest.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/306/LogHelperTest.java	2019-10-03 10:56:21.000000000 -0400
@@ -25,7 +25,8 @@
 
     @Test
     public void testGetBodyFromArchivedLog() {
-        String begin = "travis_fold:start:worker_info";
+        String
+        begin = "travis_fold:start:worker_info";
 
         String end = "Done. Your build exited with 0.";
 
