--- ../corpora/repairnator-corpus/data/70/GatherTestInformation.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/57/GatherTestInformation.java	2019-10-03 10:56:06.000000000 -0400
@@ -122,7 +122,8 @@
                                 FailureLocation failureLocation = null;
 
                                 for (FailureLocation location : this.failureLocations) {
-                                    if (location.getClassName().equals(testCase.getFullClassName())) {
+                                    if (location.getClassName()
+                                        .equals(testCase.getFullClassName())) {
                                         failureLocation = location;
                                         break;
                                     }
