--- ../corpora/repairnator-corpus/data/26/Launcher.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/473/Launcher.java	2019-10-03 10:56:30.000000000 -0400
@@ -44,7 +44,8 @@
         for (String slug : allSlugs) {
             Repository repo = RepositoryHelper.getRepositoryFromSlug(slug);
             if (repo != null) {
-                Build lastBuild = repo.getLastBuild(true);
+                Build lastBuild = repo.
+                        getLastBuild(true);
                 if (lastBuild != null) {
                     if (tool == null && language == null) {
                         result.add(repo);
