*** ../corpora/repairnator-corpus/data/41/JsonSerializer.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/48/JsonSerializer.java	2019-10-03 10:36:30.000000000 -0400
***************
*** 125,131 ****
          }
  
          result.add("testInformationPerJobId", serialize(testInformationPerJobId));
!         this.inspectors.get(inspector.getState().name()).add(result);
      }
  
      private void outputNotTestableInspector(ProjectInspector inspector) {
--- 125,132 ----
          }
  
          result.add("testInformationPerJobId", serialize(testInformationPerJobId));
!         this.inspectors.get(inspector.getState().
!                 name()).add(result);
      }
  
      private void outputNotTestableInspector(ProjectInspector inspector) {
