*** ../corpora/repairnator-corpus/data/69/NopolRepair.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/415/NopolRepair.java	2019-10-03 10:36:38.000000000 -0400
***************
*** 149,155 ****
  
              long afterNopol = new Date().getTime();
  
!             nopolInformation.setDateEnd();
  
              int localPassingTime = Math.round((afterNopol - beforeNopol) / 60000);
              nopolInformation.setPassingTime(localPassingTime);
--- 149,156 ----
  
              long afterNopol = new Date().getTime();
  
!             nopolInformation
!                 .setDateEnd();
  
              int localPassingTime = Math.round((afterNopol - beforeNopol) / 60000);
              nopolInformation.setPassingTime(localPassingTime);
