--- ../corpora/repairnator-corpus/data/69/NopolRepair.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/415/NopolRepair.java	2019-10-03 10:56:27.000000000 -0400
@@ -149,7 +149,8 @@
 
             long afterNopol = new Date().getTime();
 
-            nopolInformation.setDateEnd();
+            nopolInformation
+                .setDateEnd();
 
             int localPassingTime = Math.round((afterNopol - beforeNopol) / 60000);
             nopolInformation.setPassingTime(localPassingTime);
