*** ../corpora/repairnator-corpus/data/37/CSVSerializer4Bears.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/152/CSVSerializer4Bears.java	2019-10-03 10:36:33.000000000 -0400
***************
*** 53,59 ****
          Build previousBuild = inspector.getPreviousBuild();
          int previousBuildId = (previousBuild != null) ? previousBuild.getId() : -1;
  
!         String state = this.getPrettyPrintState(inspector.getState(), inspector.getTestInformations());
  
          String previousBuildSlug = (previousBuild != null) ? previousBuild.getRepository().getSlug() : "";
  
--- 53,60 ----
          Build previousBuild = inspector.getPreviousBuild();
          int previousBuildId = (previousBuild != null) ? previousBuild.getId() : -1;
  
!         String state = this.getPrettyPrintState(inspector.
!         getState(), inspector.getTestInformations());
  
          String previousBuildSlug = (previousBuild != null) ? previousBuild.getRepository().getSlug() : "";
  
