--- ../corpora/repairnator-corpus/data/29/GoogleSpreadSheetInspectorSerializer.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/365/GoogleSpreadSheetInspectorSerializer.java	2019-10-03 10:56:24.000000000 -0400
@@ -56,6 +56,7 @@
             dataCol.add(realState);
             dataCol.add(SerializerUtils.getHostname());
             dataCol.add(SerializerUtils.formatCompleteDate(new Date()));
+
             dataCol.add(this.getTravisUrl(build.getId(), build.getRepository().getSlug()));
             dataCol.add(typeOfFailures);
 
