*** ../corpora/repairnator-corpus/data/14/GradleLogParser.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/109/GradleLogParser.java	2019-10-03 10:36:32.000000000 -0400
***************
*** 24,30 ****
                      matcher = pattern.matcher(s);
  
                      if (matcher.matches()) {
!                         this.runningTests += Integer.parseInt(matcher.group(1));
                          this.failingTests += Integer.parseInt(matcher.group(2));
  
                          String skipped = matcher.group(4);
--- 24,31 ----
                      matcher = pattern.matcher(s);
  
                      if (matcher.matches()) {
!                         this.runningTests += Integer.
!                                 parseInt(matcher.group(1));
                          this.failingTests += Integer.parseInt(matcher.group(2));
  
                          String skipped = matcher.group(4);
