--- ../corpora/repairnator-corpus/data/48/MavenErrorHandler.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/126/MavenErrorHandler.java	2019-10-03 10:56:11.000000000 -0400
@@ -16,6 +16,7 @@
         super.consumeLine(s);
 
         this.getLogger().error(s);
-        this.inspector.addStepError(name, s);
+        this.inspector.addStepError(
+            name, s);
     }
 }
