--- ../corpora/repairnator-corpus/data/59/ComputeClasspath.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/226/ComputeClasspath.java	2019-10-03 10:56:16.000000000 -0400
@@ -82,7 +82,8 @@
 
             String[] allJars = classpathLine.split(":");
             for (String jar : allJars) {
-                File f = new File(jar);
+                File f = new File(
+                    jar);
                 this.addFileToClassPath(f);
             }
         } catch (IOException e) {
