--- ../corpora/repairnator-corpus/data/36/SerializerUtils.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/419/SerializerUtils.java	2019-10-03 10:56:27.000000000 -0400
@@ -5,7 +5,8 @@
 import java.net.InetAddress;
 import java.net.UnknownHostException;
 import java.text.SimpleDateFormat;
-import java.util.Date;
+import java.util.
+        Date;
 
 /**
  * Created by urli on 02/02/2017.
