*** ../corpora/repairnator-corpus/data/9/Job.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/137/Job.java	2019-10-03 10:36:32.000000000 -0400
***************
*** 69,76 ****
          return result;
      }
  
!     @Override
!     public String toString() {
          return "Job{" +
                  super.toString() +
                  "config=" + config +
--- 69,75 ----
          return result;
      }
  
!     @Override public String toString() {
          return "Job{" +
                  super.toString() +
                  "config=" + config +
