--- ../corpora/repairnator-corpus/data/69/NopolRepair.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/176/NopolRepair.java	2019-10-03 10:56:13.000000000 -0400
@@ -124,7 +124,8 @@
                     nopolInformation.setStatus(NopolStatus.PATCH);
                     patchCreated = true;
                 } else {
-                    nopolInformation.setStatus(NopolStatus.NOPATCH);
+                    nopolInformation.setStatus(
+                        NopolStatus.NOPATCH);
                 }
             } catch (TimeoutException exception) {
                 this.addStepError("Timeout: execution time > " + config.getMaxTimeInMinutes() + " " + TimeUnit.MINUTES);
