*** ../corpora/repairnator-corpus/data/63/ComputeSourceDir.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/387/ComputeSourceDir.java	2019-10-03 10:36:38.000000000 -0400
***************
*** 28,34 ****
  
          if (defaultSourceDir.exists()) {
              result.add(defaultSourceDir);
!             return result.toArray(new File[result.size()]);
          }
  
          this.getLogger().debug("The default source directory (" + defaultSourceDir.getPath()
--- 28,35 ----
  
          if (defaultSourceDir.exists()) {
              result.add(defaultSourceDir);
!             return result.toArray(new File[result.
!                     size()]);
          }
  
          this.getLogger().debug("The default source directory (" + defaultSourceDir.getPath()
