--- ../corpora/repairnator-corpus/data/47/MavenFilterTestOutputHandler.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/159/MavenFilterTestOutputHandler.java	2019-10-03 10:56:12.000000000 -0400
@@ -71,7 +71,8 @@
 
         if (matcher.matches()) {
             this.runningTests += Integer.parseInt(matcher.group(1));
-            int failing = Integer.parseInt(matcher.group(2));
+            int failing = Integer.parseInt(matcher
+                .group(2));
 
             this.failingTests += failing;
             this.erroringTests += Integer.parseInt(matcher.group(3));
