*** ../corpora/repairnator-corpus/data/69/NopolRepair.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/166/NopolRepair.java	2019-10-03 10:36:33.000000000 -0400
***************
*** 81,87 ****
                  timeout = MIN_TIMEOUT;
              }
  
!             nopolInformation.setAllocatedTime(timeout);
  
              this.getLogger().debug("Launching repair with Nopol for following test class: " + testClass
                      + " (should timeout in " + timeout + " minutes)");
--- 81,88 ----
                  timeout = MIN_TIMEOUT;
              }
  
!             nopolInformation.
!             setAllocatedTime(timeout);
  
              this.getLogger().debug("Launching repair with Nopol for following test class: " + testClass
                      + " (should timeout in " + timeout + " minutes)");
