*** ../corpora/repairnator-corpus/data/40/CustomExclusionStrategy.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/392/CustomExclusionStrategy.java	2019-10-03 10:36:38.000000000 -0400
***************
*** 9,15 ****
   * Created by urli on 11/01/2017.
   */
  public class CustomExclusionStrategy implements ExclusionStrategy {
!     public boolean shouldSkipField(FieldAttributes fieldAttributes) {
          return fieldAttributes.getName().equals("lastBuild") || fieldAttributes.getName().equals("logger")
                  || fieldAttributes.getDeclaredClass().equals(Class.class)
                  || fieldAttributes.hasModifier(Modifier.PROTECTED);
--- 9,16 ----
   * Created by urli on 11/01/2017.
   */
  public class CustomExclusionStrategy implements ExclusionStrategy {
!     public boolean shouldSkipField(FieldAttributes
!     fieldAttributes) {
          return fieldAttributes.getName().equals("lastBuild") || fieldAttributes.getName().equals("logger")
                  || fieldAttributes.getDeclaredClass().equals(Class.class)
                  || fieldAttributes.hasModifier(Modifier.PROTECTED);
