*** ../corpora/repairnator-corpus/data/47/MavenFilterTestOutputHandler.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/267/MavenFilterTestOutputHandler.java	2019-10-03 10:36:35.000000000 -0400
***************
*** 3,9 ****
  import fr.inria.spirals.repairnator.process.inspectors.ProjectInspector;
  
  import java.util.ArrayList;
! import java.util.List;
  import java.util.regex.Matcher;
  import java.util.regex.Pattern;
  
--- 3,10 ----
  import fr.inria.spirals.repairnator.process.inspectors.ProjectInspector;
  
  import java.util.ArrayList;
! import java.util.
!     List;
  import java.util.regex.Matcher;
  import java.util.regex.Pattern;
  
