--- ../corpora/repairnator-corpus/data/49/MavenMuteOutputHandler.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/211/MavenMuteOutputHandler.java	2019-10-03 10:56:15.000000000 -0400
@@ -8,7 +8,7 @@
 public class MavenMuteOutputHandler extends MavenOutputHandler {
     public MavenMuteOutputHandler(ProjectInspector inspector, String name) {
         super(inspector, name);
-    }
+}
 
     @Override
     public void consumeLine(String s) {
