--- ../corpora/repairnator-corpus/data/71/AbstractStep.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/31/AbstractStep.java	2019-10-03 10:56:05.000000000 -0400
@@ -114,7 +114,8 @@
         }
     }
 
-    private void serializeData() {
+    private void serializeData()
+    {
         if (serializers != null) {
             for (AbstractDataSerializer serializer : this.serializers) {
                 serializer.serializeData(this.inspector);
