--- ../corpora/repairnator-corpus/data/64/BuildShouldFail.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/55/BuildShouldFail.java	2019-10-03 10:56:06.000000000 -0400
@@ -14,7 +14,8 @@
         ProjectInspector inspector = gatherTestInformation.getInspector();
         if (gatherTestInformation.getState() == ProjectState.HASTESTFAILURE) {
             inspector.setReproducedAsFail(true);
-            if (inspector.isAboutAPreviousBuild()) {
+            if (
+                inspector.isAboutAPreviousBuild()) {
                 if (inspector.getPreviousBuild().getBuildStatus() == BuildStatus.FAILED) {
                     // So, 1) the current passing build can be reproduced and 2)
                     // its previous build is a failing build with failing tests
