*** ../corpora/repairnator-corpus/data/64/BuildShouldFail.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/55/BuildShouldFail.java	2019-10-03 10:36:31.000000000 -0400
***************
*** 14,20 ****
          ProjectInspector inspector = gatherTestInformation.getInspector();
          if (gatherTestInformation.getState() == ProjectState.HASTESTFAILURE) {
              inspector.setReproducedAsFail(true);
!             if (inspector.isAboutAPreviousBuild()) {
                  if (inspector.getPreviousBuild().getBuildStatus() == BuildStatus.FAILED) {
                      // So, 1) the current passing build can be reproduced and 2)
                      // its previous build is a failing build with failing tests
--- 14,21 ----
          ProjectInspector inspector = gatherTestInformation.getInspector();
          if (gatherTestInformation.getState() == ProjectState.HASTESTFAILURE) {
              inspector.setReproducedAsFail(true);
!             if (
!                 inspector.isAboutAPreviousBuild()) {
                  if (inspector.getPreviousBuild().getBuildStatus() == BuildStatus.FAILED) {
                      // So, 1) the current passing build can be reproduced and 2)
                      // its previous build is a failing build with failing tests
