--- ../corpora/repairnator-corpus/data/24/PRInformationHelperTest.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/417/PRInformationHelperTest.java	2019-10-03 10:56:27.000000000 -0400
@@ -36,7 +36,8 @@
         expectedConfig.setLanguage("java");
         expectedBuild.setConfig(expectedConfig);
 
-        Commit commit = new Commit();
+        Commit commit
+            = new Commit();
         commit.setSha("982bbbc0697b063bf04306a45fe087a5aec3e1f0");
         commit.setBranch("master");
         commit.setCommittedAt(TestUtils.getDate(2016,12,27,17,59,07));
