--- ../corpora/repairnator-corpus/data/72/CheckoutPreviousBuild.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/355/CheckoutPreviousBuild.java	2019-10-03 10:56:23.000000000 -0400
@@ -19,7 +19,8 @@
     }
 
     protected void businessExecute() {
-        this.getLogger().debug("Checking out previous build...");
+        this.getLogger()
+        .debug("Checking out previous build...");
 
         this.inspector.setPreviousBuildFlag(true);
 
