--- ../corpora/repairnator-corpus/data/66/CloneRepository.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/36/CloneRepository.java	2019-10-03 10:56:05.000000000 -0400
@@ -21,7 +21,8 @@
 import org.eclipse.jgit.lib.ObjectId;
 import org.eclipse.jgit.revwalk.RevCommit;
 import org.eclipse.jgit.revwalk.RevWalk;
-import org.eclipse.jgit.transport.URIish;
+import org.eclipse.jgit.transport.
+        URIish;
 import org.kohsuke.github.GHCommit;
 import org.kohsuke.github.GHCompare;
 import org.kohsuke.github.GHRateLimit;
