*** ../corpora/repairnator-corpus/data/65/ContractForGatherTestInformation.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/199/ContractForGatherTestInformation.java	2019-10-03 10:36:34.000000000 -0400
***************
*** 1,4 ****
! package fr.inria.spirals.repairnator.process.step;
  
  /**
   * Created by fermadeiral.
--- 1,5 ----
! package fr.inria.spirals.
! repairnator.process.step;
  
  /**
   * Created by fermadeiral.
