--- ../corpora/repairnator-corpus/data/69/NopolRepair.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/385/NopolRepair.java	2019-10-03 10:56:25.000000000 -0400
@@ -141,7 +141,8 @@
 
             try {
                 Runtime.getRuntime().exec(CMD_KILL_GZOLTAR_AGENT);
-            } catch (IOException e) {
+            } catch (IOException e)
+            {
                 this.getLogger()
                         .error("Error while killing gzoltar agent using following command: " + CMD_KILL_GZOLTAR_AGENT);
                 this.getLogger().error(e.getMessage());
