--- ../corpora/repairnator-corpus/data/39/CSVSerializer4RepairNator.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/367/CSVSerializer4RepairNator.java	2019-10-03 10:56:24.000000000 -0400
@@ -45,7 +45,8 @@
 
         String realState = (inspector.getState() != null) ? inspector.getState().name() : "null";
 
-        this.writeData(build.getId(), build.getRepository().getSlug(), state, realState, build.getPullRequestNumber(),
+        this.writeData(build.getId(),
+        build.getRepository().getSlug(), state, realState, build.getPullRequestNumber(),
                 build.getFinishedAt());
     }
 }
