--- ../corpora/repairnator-corpus/data/66/CloneRepository.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/46/CloneRepository.java	2019-10-03 10:56:06.000000000 -0400
@@ -234,7 +234,8 @@
                 this.writeProperty("pr-id", this.build.getPullRequestNumber() + "");
 
                 this.getLogger()
-                        .debug("Reproduce the PR for " + repository + " by fetching remote branch and merging.");
+                        .debug("Reproduce the PR for " + repository + " by fetching remote branch and merging."
+                    );
                 String remoteBranchPath = GITHUB_ROOT_REPO + prInformation.getOtherRepo().getSlug() + ".git";
 
                 RemoteAddCommand remoteBranchCommand = git.remoteAdd();
