--- ../corpora/repairnator-corpus/data/11/MavenLogParser.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/460/MavenLogParser.java	2019-10-03 10:56:29.000000000 -0400
@@ -38,7 +38,7 @@
             if (nbMatch == 5) {
                 this.skippingTests += Integer.parseInt(mvnTestNumberMatcher.group(5));
             }
-        }
+          }
     }
 
     @Override
