*** ../corpora/repairnator-corpus/data/72/CheckoutPreviousBuild.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/315/CheckoutPreviousBuild.java	2019-10-03 10:36:36.000000000 -0400
***************
*** 33,39 ****
                          .debug("Get the commit " + commitCheckout + " for repo " + this.inspector.getRepoSlug());
                  git.checkout().setName(commitCheckout).call();
                  this.state = ProjectState.PREVIOUSBUILDCHECKEDOUT;
!             } else {
                  this.addStepError("Error while getting the commit to checkout from the repo.");
                  this.shouldStop = true;
                  this.state = ProjectState.PREVIOUSBUILDNOTCHECKEDOUT;
--- 33,40 ----
                          .debug("Get the commit " + commitCheckout + " for repo " + this.inspector.getRepoSlug());
                  git.checkout().setName(commitCheckout).call();
                  this.state = ProjectState.PREVIOUSBUILDCHECKEDOUT;
!             }
!             else {
                  this.addStepError("Error while getting the commit to checkout from the repo.");
                  this.shouldStop = true;
                  this.state = ProjectState.PREVIOUSBUILDNOTCHECKEDOUT;
