*** ../corpora/repairnator-corpus/data/62/CheckoutSourceCodeForPreviousBuild.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/38/CheckoutSourceCodeForPreviousBuild.java	2019-10-03 10:36:30.000000000 -0400
***************
*** 22,27 ****
--- 22,28 ----
          this.getLogger().debug("Checking out the source code folder for previous build...");
  
          Git git;
+ 
          try {
              git = Git.open(new File(inspector.getRepoLocalPath()));
              String commitCheckout = build.getCommit().getSha();
