--- ../corpora/repairnator-corpus/data/62/CheckoutSourceCodeForPreviousBuild.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/38/CheckoutSourceCodeForPreviousBuild.java	2019-10-03 10:56:05.000000000 -0400
@@ -22,6 +22,7 @@
         this.getLogger().debug("Checking out the source code folder for previous build...");
 
         Git git;
+
         try {
             git = Git.open(new File(inspector.getRepoLocalPath()));
             String commitCheckout = build.getCommit().getSha();
