--- ../corpora/repairnator-corpus/data/9/Job.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/124/Job.java	2019-10-03 10:56:11.000000000 -0400
@@ -65,7 +65,8 @@
     public int hashCode() {
         int result = super.hashCode();
         result = 31 * result + (config != null ? config.hashCode() : 0);
-        result = 31 * result + (log != null ? log.hashCode() : 0);
+        result = 31 * result + (log != null ?
+            log.hashCode() : 0);
         return result;
     }
 
