--- ../corpora/repairnator-corpus/data/37/CSVSerializer4Bears.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/499/CSVSerializer4Bears.java	2019-10-03 10:56:31.000000000 -0400
@@ -3,7 +3,8 @@
 import java.io.BufferedWriter;
 import java.util.Date;
 
-import org.slf4j.Logger;
+import org
+    .slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import fr.inria.spirals.jtravis.entities.Build;
