*** ../corpora/repairnator-corpus/data/44/ProjectInspector4Bears.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/425/ProjectInspector4Bears.java	2019-10-03 10:36:38.000000000 -0400
***************
*** 46,52 ****
  
              lastStep = gatherTestInformation;
          } else {
!             if (this.buildToBeInspected.getStatus() == ScannedBuildStatus.PASSING_AND_PASSING) {
                  AbstractStep gatherTestInformation = new GatherTestInformation(this, new BuildShouldPass());
                  AbstractStep checkoutSourceCodeForPreviousBuild = new CheckoutSourceCodeForPreviousBuild(this);
                  AbstractStep buildRepoForPreviousBuild2 = new BuildProject(this);
--- 46,52 ----
  
              lastStep = gatherTestInformation;
          } else {
!           if (this.buildToBeInspected.getStatus() == ScannedBuildStatus.PASSING_AND_PASSING) {
                  AbstractStep gatherTestInformation = new GatherTestInformation(this, new BuildShouldPass());
                  AbstractStep checkoutSourceCodeForPreviousBuild = new CheckoutSourceCodeForPreviousBuild(this);
                  AbstractStep buildRepoForPreviousBuild2 = new BuildProject(this);
