*** ../corpora/repairnator-corpus/data/34/GoogleSpreadSheetScannerSerializer.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/16/GoogleSpreadSheetScannerSerializer.java	2019-10-03 10:36:30.000000000 -0400
***************
*** 33,39 ****
          if (this.sheets != null) {
              List<Object> dataCol = new ArrayList<Object>();
              dataCol.add(SerializerUtils.getHostname());
!             dataCol.add(SerializerUtils.formatCompleteDate(new Date()));
              dataCol.add(SerializerUtils.formatCompleteDate(this.scanner.getLimitDate()));
              dataCol.add(this.scanner.getTotalRepoNumber());
              dataCol.add(this.scanner.getTotalRepoUsingTravis());
--- 33,40 ----
          if (this.sheets != null) {
              List<Object> dataCol = new ArrayList<Object>();
              dataCol.add(SerializerUtils.getHostname());
!             dataCol
!                     .add(SerializerUtils.formatCompleteDate(new Date()));
              dataCol.add(SerializerUtils.formatCompleteDate(this.scanner.getLimitDate()));
              dataCol.add(this.scanner.getTotalRepoNumber());
              dataCol.add(this.scanner.getTotalRepoUsingTravis());
