--- ../corpora/repairnator-corpus/data/42/AbstractDataSerializer.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/296/AbstractDataSerializer.java	2019-10-03 10:56:21.000000000 -0400
@@ -18,7 +18,8 @@
         return TRAVIS_URL + slug + "/builds/" + buildId;
     }
 
-    protected String getPrettyPrintState(ProjectState state, GatherTestInformation gatherTestInformation) {
+    protected String getPrettyPrintState(
+        ProjectState state, GatherTestInformation gatherTestInformation) {
         switch (state) {
             case NONE:
             case INIT:
