--- ../corpora/repairnator-corpus/data/25/LogHelperTest.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/281/LogHelperTest.java	2019-10-03 10:56:20.000000000 -0400
@@ -31,7 +31,8 @@
 
         int length = 1499914;
 
-        Log obtainedLog = LogHelper.getLogFromId(135819715);
+        Log obtainedLog =
+            LogHelper.getLogFromId(135819715);
         String body = obtainedLog.getBody();
         body = body.trim();
 
