*** ../corpora/repairnator-corpus/data/74/ProjectScanner.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/6/ProjectScanner.java	2019-10-03 10:36:29.000000000 -0400
***************
*** 158,164 ****
          List<String> result = new ArrayList<String>();
          File file = new File(path);
          BufferedReader reader = new BufferedReader(new FileReader(file));
!         while (reader.ready()) {
              result.add(reader.readLine().trim());
          }
          return result;
--- 158,165 ----
          List<String> result = new ArrayList<String>();
          File file = new File(path);
          BufferedReader reader = new BufferedReader(new FileReader(file));
!         while (reader.ready())
!         {
              result.add(reader.readLine().trim());
          }
          return result;
