*** ../corpora/repairnator-corpus/data/62/CheckoutSourceCodeForPreviousBuild.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/454/CheckoutSourceCodeForPreviousBuild.java	2019-10-03 10:36:39.000000000 -0400
***************
*** 34,40 ****
                          .call();
                  this.state = ProjectState.PREVIOUSBUILDCODECHECKEDOUT;
              } else {
!                 this.addStepError("Error while getting the commit to checkout from the repo.");
                  this.shouldStop = true;
                  this.state = ProjectState.PREVIOUSBUILDCODENOTCHECKEDOUT;
                  return;
--- 34,41 ----
                          .call();
                  this.state = ProjectState.PREVIOUSBUILDCODECHECKEDOUT;
              } else {
!                 this.
!                 addStepError("Error while getting the commit to checkout from the repo.");
                  this.shouldStop = true;
                  this.state = ProjectState.PREVIOUSBUILDCODENOTCHECKEDOUT;
                  return;
