*** ../corpora/repairnator-corpus/data/33/GoogleSpreadSheetEndProcessSerializer.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/451/GoogleSpreadSheetEndProcessSerializer.java	2019-10-03 10:36:39.000000000 -0400
***************
*** 48,54 ****
  
              List<Object> dataCol = new ArrayList<Object>();
              dataCol.add(SerializerUtils.getHostname());
!             dataCol.add(SerializerUtils.formatOnlyDay(this.beginDate));
              dataCol.add(SerializerUtils.formatCompleteDate(this.beginDate));
              dataCol.add(SerializerUtils.formatCompleteDate(now));
              dataCol.add(humanDuration);
--- 48,55 ----
  
              List<Object> dataCol = new ArrayList<Object>();
              dataCol.add(SerializerUtils.getHostname());
! 
!     dataCol.add(SerializerUtils.formatOnlyDay(this.beginDate));
              dataCol.add(SerializerUtils.formatCompleteDate(this.beginDate));
              dataCol.add(SerializerUtils.formatCompleteDate(now));
              dataCol.add(humanDuration);
