*** ../corpora/repairnator-corpus/data/17/ConfigHelper.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/99/ConfigHelper.java	2019-10-03 10:36:32.000000000 -0400
***************
*** 9,15 ****
   * @author Simon Urli
   */
  public class ConfigHelper extends AbstractHelper {
!     public static Config getConfigFromJsonElement(JsonElement jsonConfig) {
          return createGson().fromJson(jsonConfig, Config.class);
      }
  }
--- 9,16 ----
   * @author Simon Urli
   */
  public class ConfigHelper extends AbstractHelper {
!     public static Config getConfigFromJsonElement(JsonElement
!     jsonConfig) {
          return createGson().fromJson(jsonConfig, Config.class);
      }
  }
