*** ../corpora/repairnator-corpus/data/57/BuildToBeInspected.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/272/BuildToBeInspected.java	2019-10-03 10:36:35.000000000 -0400
***************
*** 1,6 ****
  package fr.inria.spirals.repairnator.process;
  
! import fr.inria.spirals.jtravis.entities.Build;
  
  /**
   * Created by fernanda on 24/02/17.
--- 1,7 ----
  package fr.inria.spirals.repairnator.process;
  
! import fr.inria.spirals.
!         jtravis.entities.Build;
  
  /**
   * Created by fernanda on 24/02/17.
