*** ../corpora/repairnator-corpus/data/55/ComparatorFailureLocation.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/368/ComparatorFailureLocation.java	2019-10-03 10:36:37.000000000 -0400
***************
*** 14,20 ****
              int diffFail = o2.getNbFailures() - o1.getNbFailures();
              int diffErr = o2.getNbErrors() - o1.getNbErrors();
  
!             return (diffFail != 0) ? diffFail : diffErr;
          }
      }
  }
--- 14,21 ----
              int diffFail = o2.getNbFailures() - o1.getNbFailures();
              int diffErr = o2.getNbErrors() - o1.getNbErrors();
  
!             return (diffFail != 0)
!                 ? diffFail : diffErr;
          }
      }
  }
