--- ../corpora/repairnator-corpus/data/67/BuildProject.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/397/BuildProject.java	2019-10-03 10:56:25.000000000 -0400
@@ -24,7 +24,8 @@
         MavenHelper helper = new MavenHelper(this.getPom(), "install", properties, this.getClass().getSimpleName(),
                 this.inspector, true);
 
-        int result = helper.run();
+        int result
+            = helper.run();
 
         if (result == MavenHelper.MAVEN_SUCCESS) {
             this.state = ProjectState.BUILDABLE;
