--- ../corpora/repairnator-corpus/data/1/JobPojo.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/12/JobPojo.java	2019-10-03 10:56:03.000000000 -0400
@@ -154,8 +154,7 @@
                 ", logId=" + logId +
                 ", state='" + state + '\'' +
                 ", number='" + number + '\'' +
-                ", startedAt=" + startedAt +
-                ", finishedAt=" + finishedAt +
+                ", startedAt=" + startedAt + ", finishedAt=" + finishedAt +
                 ", queue='" + queue + '\'' +
                 ", allowFailure=" + allowFailure +
                 '}';
