*** ../corpora/repairnator-corpus/data/1/JobPojo.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/12/JobPojo.java	2019-10-03 10:36:30.000000000 -0400
***************
*** 154,161 ****
                  ", logId=" + logId +
                  ", state='" + state + '\'' +
                  ", number='" + number + '\'' +
!                 ", startedAt=" + startedAt +
!                 ", finishedAt=" + finishedAt +
                  ", queue='" + queue + '\'' +
                  ", allowFailure=" + allowFailure +
                  '}';
--- 154,160 ----
                  ", logId=" + logId +
                  ", state='" + state + '\'' +
                  ", number='" + number + '\'' +
!                 ", startedAt=" + startedAt + ", finishedAt=" + finishedAt +
                  ", queue='" + queue + '\'' +
                  ", allowFailure=" + allowFailure +
                  '}';
