--- ../corpora/repairnator-corpus/data/28/Launcher.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/45/Launcher.java	2019-10-03 10:56:06.000000000 -0400
@@ -388,7 +388,8 @@
     }
 
     private List<ProjectInspector> cloneAndRepair(List<BuildToBeInspected> buildsToBeInspected, String workspace) throws IOException {
-        if (mode != RepairMode.NOPOLONLY) {
+        if (mode != RepairMode.NOPOLONLY
+    ) {
             initWorkspace(workspace);
         }
 
