*** ../corpora/repairnator-corpus/data/28/Launcher.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/45/Launcher.java	2019-10-03 10:36:30.000000000 -0400
***************
*** 388,394 ****
      }
  
      private List<ProjectInspector> cloneAndRepair(List<BuildToBeInspected> buildsToBeInspected, String workspace) throws IOException {
!         if (mode != RepairMode.NOPOLONLY) {
              initWorkspace(workspace);
          }
  
--- 388,395 ----
      }
  
      private List<ProjectInspector> cloneAndRepair(List<BuildToBeInspected> buildsToBeInspected, String workspace) throws IOException {
!         if (mode != RepairMode.NOPOLONLY
!     ) {
              initWorkspace(workspace);
          }
  
