--- ../corpora/repairnator-corpus/data/66/CloneRepository.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/54/CloneRepository.java	2019-10-03 10:56:06.000000000 -0400
@@ -52,7 +52,8 @@
 
     private void showGitHubRateInformation(GitHub gh) throws IOException {
         GHRateLimit rateLimit = gh.getRateLimit();
-        SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm:ss");
+        SimpleDateFormat dateFormat
+            = new SimpleDateFormat("HH:mm:ss");
         this.getLogger().info("GitHub ratelimit: Limit: " + rateLimit.limit + " Remaining: " + rateLimit.remaining
                 + " Reset hour: " + dateFormat.format(rateLimit.reset));
     }
