*** ../corpora/repairnator-corpus/data/25/LogHelperTest.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/416/LogHelperTest.java	2019-10-03 10:36:38.000000000 -0400
***************
*** 15,21 ****
      public void testGetLogFromId() {
          Log expectedLog = new Log();
          expectedLog.setId(135819715);
!         expectedLog.setJobId(185719844);
          expectedLog.setType("Log");
          expectedLog.setBody("");
  
--- 15,22 ----
      public void testGetLogFromId() {
          Log expectedLog = new Log();
          expectedLog.setId(135819715);
!         expectedLog.
!         setJobId(185719844);
          expectedLog.setType("Log");
          expectedLog.setBody("");
  
