--- ../corpora/repairnator-corpus/data/48/MavenErrorHandler.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/63/MavenErrorHandler.java	2019-10-03 10:56:07.000000000 -0400
@@ -11,8 +11,7 @@
         super(inspector, name);
     }
 
-    @Override
-    public void consumeLine(String s) {
+    @Override public void consumeLine(String s) {
         super.consumeLine(s);
 
         this.getLogger().error(s);
