*** ../corpora/repairnator-corpus/data/68/PushIncriminatedBuild.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/359/PushIncriminatedBuild.java	2019-10-03 10:36:37.000000000 -0400
***************
*** 7,13 ****
  import org.eclipse.jgit.api.RemoteAddCommand;
  import org.eclipse.jgit.api.errors.GitAPIException;
  import org.eclipse.jgit.lib.ObjectId;
! import org.eclipse.jgit.lib.PersonIdent;
  import org.eclipse.jgit.lib.Ref;
  import org.eclipse.jgit.transport.CredentialsProvider;
  import org.eclipse.jgit.transport.URIish;
--- 7,14 ----
  import org.eclipse.jgit.api.RemoteAddCommand;
  import org.eclipse.jgit.api.errors.GitAPIException;
  import org.eclipse.jgit.lib.ObjectId;
! import org
!     .eclipse.jgit.lib.PersonIdent;
  import org.eclipse.jgit.lib.Ref;
  import org.eclipse.jgit.transport.CredentialsProvider;
  import org.eclipse.jgit.transport.URIish;
