--- ../corpora/repairnator-corpus/data/24/PRInformationHelperTest.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/203/PRInformationHelperTest.java	2019-10-03 10:56:15.000000000 -0400
@@ -26,7 +26,8 @@
         expectedBuild.setPullRequestTitle("Typo and wording fixes");
         expectedBuild.setPullRequestNumber(1079);
         expectedBuild.setJobIds(Arrays.asList(new Integer[]{187029371}));
-        expectedBuild.setDuration(462);
+        expectedBuild.setDuration(
+            462);
         expectedBuild.setState("passed");
         expectedBuild.setStartedAt(TestUtils.getDate(2016,12,27,18,02,53));
         expectedBuild.setFinishedAt(TestUtils.getDate(2016,12,27,18,10,35));
