*** ../corpora/repairnator-corpus/data/40/CustomExclusionStrategy.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/440/CustomExclusionStrategy.java	2019-10-03 10:36:39.000000000 -0400
***************
*** 12,18 ****
      public boolean shouldSkipField(FieldAttributes fieldAttributes) {
          return fieldAttributes.getName().equals("lastBuild") || fieldAttributes.getName().equals("logger")
                  || fieldAttributes.getDeclaredClass().equals(Class.class)
!                 || fieldAttributes.hasModifier(Modifier.PROTECTED);
      }
  
      public boolean shouldSkipClass(Class<?> aClass) {
--- 12,19 ----
      public boolean shouldSkipField(FieldAttributes fieldAttributes) {
          return fieldAttributes.getName().equals("lastBuild") || fieldAttributes.getName().equals("logger")
                  || fieldAttributes.getDeclaredClass().equals(Class.class)
!                 ||
!                 fieldAttributes.hasModifier(Modifier.PROTECTED);
      }
  
      public boolean shouldSkipClass(Class<?> aClass) {
