*** ../corpora/repairnator-corpus/data/36/SerializerUtils.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/120/SerializerUtils.java	2019-10-03 10:36:32.000000000 -0400
***************
*** 27,33 ****
          try {
              return InetAddress.getLocalHost().getHostName();
  
!         } catch (UnknownHostException e) {
              return "unknown host";
          }
      }
--- 27,34 ----
          try {
              return InetAddress.getLocalHost().getHostName();
  
!         } catch (UnknownHostException e)
!         {
              return "unknown host";
          }
      }
