--- ../corpora/repairnator-corpus/data/54/FailureType.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/308/FailureType.java	2019-10-03 10:56:21.000000000 -0400
@@ -41,7 +41,8 @@
         if (isError != that.isError) {
             return false;
         }
-        if (failureName != null ? !failureName.equals(that.failureName) : that.failureName != null) {
+        if (failureName != null ? !failureName.equals(that.failureName) : that.failureName != null
+    ) {
             return false;
         }
         return failureDetail != null ? failureDetail.equals(that.failureDetail) : that.failureDetail == null;
