*** ../corpora/repairnator-corpus/data/70/GatherTestInformation.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/268/GatherTestInformation.java	2019-10-03 10:36:35.000000000 -0400
***************
*** 11,17 ****
  import java.util.Arrays;
  import java.util.HashSet;
  import java.util.List;
! import java.util.Locale;
  import java.util.Set;
  
  import org.apache.commons.lang3.StringUtils;
--- 11,18 ----
  import java.util.Arrays;
  import java.util.HashSet;
  import java.util.List;
! import java.util
!     .Locale;
  import java.util.Set;
  
  import org.apache.commons.lang3.StringUtils;
