*** ../corpora/repairnator-corpus/data/25/LogHelperTest.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/430/LogHelperTest.java	2019-10-03 10:36:39.000000000 -0400
***************
*** 14,20 ****
      @Test
      public void testGetLogFromId() {
          Log expectedLog = new Log();
!         expectedLog.setId(135819715);
          expectedLog.setJobId(185719844);
          expectedLog.setType("Log");
          expectedLog.setBody("");
--- 14,20 ----
      @Test
      public void testGetLogFromId() {
          Log expectedLog = new Log();
!     expectedLog.setId(135819715);
          expectedLog.setJobId(185719844);
          expectedLog.setType("Log");
          expectedLog.setBody("");
