*** ../corpora/repairnator-corpus/data/54/FailureType.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/332/FailureType.java	2019-10-03 10:36:37.000000000 -0400
***************
*** 50,56 ****
      @Override
      public int hashCode() {
          int result = failureName != null ? failureName.hashCode() : 0;
!         result = 31 * result + (failureDetail != null ? failureDetail.hashCode() : 0);
          result = 31 * result + (isError ? 1 : 0);
          return result;
      }
--- 50,57 ----
      @Override
      public int hashCode() {
          int result = failureName != null ? failureName.hashCode() : 0;
!         result = 31 * result + (failureDetail != null ? failureDetail
!             .hashCode() : 0);
          result = 31 * result + (isError ? 1 : 0);
          return result;
      }
