--- ../corpora/repairnator-corpus/data/54/FailureType.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/332/FailureType.java	2019-10-03 10:56:22.000000000 -0400
@@ -50,7 +50,8 @@
     @Override
     public int hashCode() {
         int result = failureName != null ? failureName.hashCode() : 0;
-        result = 31 * result + (failureDetail != null ? failureDetail.hashCode() : 0);
+        result = 31 * result + (failureDetail != null ? failureDetail
+            .hashCode() : 0);
         result = 31 * result + (isError ? 1 : 0);
         return result;
     }
