--- ../corpora/repairnator-corpus/data/35/GoogleSpreadSheetFactory.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/150/GoogleSpreadSheetFactory.java	2019-10-03 10:56:12.000000000 -0400
@@ -86,7 +86,6 @@
                 clientSecrets, this.scopes).setDataStoreFactory(this.dataStoreFactory).build();
         Credential credential = new AuthorizationCodeInstalledApp(flow, new LocalServerReceiver()).authorize("user");
         this.logger.debug("Credentials saved to " + DATA_STORE_DIR.getAbsolutePath());
-
         this.sheets = new Sheets.Builder(this.httpTransport, this.jsonFactory, credential)
                 .setApplicationName(APPLICATION_NAME).build();
     }
