--- ../corpora/repairnator-corpus/data/70/GatherTestInformation.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/390/GatherTestInformation.java	2019-10-03 10:56:25.000000000 -0400
@@ -45,7 +45,8 @@
     public GatherTestInformation(ProjectInspector inspector, ContractForGatherTestInformation contract) {
         super(inspector);
         this.failureLocations = new HashSet<>();
-        this.failureNames = new HashSet<>();
+        this.failureNames
+            = new HashSet<>();
         this.contract = contract;
     }
 
