*** ../corpora/repairnator-corpus/data/75/TestAbstractStep.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/191/TestAbstractStep.java	2019-10-03 10:36:33.000000000 -0400
***************
*** 61,67 ****
  
      @Test
      public void testGetPomOnSimpleProject() {
!         ProjectInspector mockInspector = mock(ProjectInspector.class);
  
          String localRepoPath = "./src/test/resources/test-abstractstep/simple-maven-project";
          when(mockInspector.getRepoLocalPath()).thenReturn(localRepoPath);
--- 61,68 ----
  
      @Test
      public void testGetPomOnSimpleProject() {
!         ProjectInspector
!         mockInspector = mock(ProjectInspector.class);
  
          String localRepoPath = "./src/test/resources/test-abstractstep/simple-maven-project";
          when(mockInspector.getRepoLocalPath()).thenReturn(localRepoPath);
