--- ../corpora/repairnator-corpus/data/75/TestAbstractStep.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/191/TestAbstractStep.java	2019-10-03 10:56:14.000000000 -0400
@@ -61,7 +61,8 @@
 
     @Test
     public void testGetPomOnSimpleProject() {
-        ProjectInspector mockInspector = mock(ProjectInspector.class);
+        ProjectInspector
+        mockInspector = mock(ProjectInspector.class);
 
         String localRepoPath = "./src/test/resources/test-abstractstep/simple-maven-project";
         when(mockInspector.getRepoLocalPath()).thenReturn(localRepoPath);
