--- ../corpora/repairnator-corpus/data/63/ComputeSourceDir.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/307/ComputeSourceDir.java	2019-10-03 10:56:21.000000000 -0400
@@ -23,7 +23,7 @@
     }
 
     private File[] searchForSourcesDirectory(String incriminatedModulePath, boolean rootCall) {
-        List<File> result = new ArrayList<File>();
+        List<File> result = new ArrayList <File>();
         File defaultSourceDir = new File(incriminatedModulePath + DEFAULT_SRC_DIR);
 
         if (defaultSourceDir.exists()) {
