--- ../corpora/repairnator-corpus/data/75/TestAbstractStep.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/4/TestAbstractStep.java	2019-10-03 10:56:03.000000000 -0400
@@ -86,7 +86,8 @@
 
         // return this path but set the flag to stop
         assertThat(step1.getPom(), is(expectedPomPath));
-        assertThat(step1.shouldStop, is(true));
+        assertThat(step1.shouldStop,
+        is(true));
     }
 
     @Test
