*** ../corpora/repairnator-corpus/data/55/ComparatorFailureLocation.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/410/ComparatorFailureLocation.java	2019-10-03 10:36:38.000000000 -0400
***************
*** 5,11 ****
  /**
   * Comparator to order FailureLocation based on the number of failing tests
   */
! public class ComparatorFailureLocation implements Comparator<FailureLocation> {
      @Override
      public int compare(FailureLocation o1, FailureLocation o2) {
          if (o1.equals(o2)) {
--- 5,11 ----
  /**
   * Comparator to order FailureLocation based on the number of failing tests
   */
! public class ComparatorFailureLocation implements Comparator< FailureLocation> {
      @Override
      public int compare(FailureLocation o1, FailureLocation o2) {
          if (o1.equals(o2)) {
