*** ../corpora/repairnator-corpus/data/32/GoogleSpreadSheetInspectorSerializer4Bears.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/52/GoogleSpreadSheetInspectorSerializer4Bears.java	2019-10-03 10:36:31.000000000 -0400
***************
*** 66,72 ****
              try {
                  AppendValuesResponse response = this.sheets.spreadsheets().values()
                          .append(GoogleSpreadSheetFactory.getSpreadsheetID(), RANGE, valueRange)
!                         .setInsertDataOption("INSERT_ROWS").setValueInputOption("USER_ENTERED").execute();
                  if (response != null && response.getUpdates().getUpdatedCells() > 0) {
                      this.logger.debug("Data have been inserted in Google Spreadsheet.");
                  }
--- 66,73 ----
              try {
                  AppendValuesResponse response = this.sheets.spreadsheets().values()
                          .append(GoogleSpreadSheetFactory.getSpreadsheetID(), RANGE, valueRange)
!                         .setInsertDataOption("INSERT_ROWS").setValueInputOption(
!                             "USER_ENTERED").execute();
                  if (response != null && response.getUpdates().getUpdatedCells() > 0) {
                      this.logger.debug("Data have been inserted in Google Spreadsheet.");
                  }
