--- ../corpora/repairnator-corpus/data/69/NopolRepair.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/131/NopolRepair.java	2019-10-03 10:56:11.000000000 -0400
@@ -95,7 +95,8 @@
             config.setLocalizer(Config.NopolLocalizer.GZOLTAR);
             config.setSolverPath(this.inspector.getNopolSolverPath());
             config.setSynthesis(Config.NopolSynthesis.DYNAMOTH);
-            config.setType(StatementType.PRE_THEN_COND);
+            config
+                    .setType(StatementType.PRE_THEN_COND);
             config.setProjectSourcePath(sourcesStr);
 
             nopolInformation.setConfig(config);
