*** ../corpora/repairnator-corpus/data/35/GoogleSpreadSheetFactory.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/252/GoogleSpreadSheetFactory.java	2019-10-03 10:36:35.000000000 -0400
***************
*** 1,7 ****
  package fr.inria.spirals.repairnator.serializer.gsheet;
  
  import com.google.api.client.auth.oauth2.Credential;
! import com.google.api.client.extensions.java6.auth.oauth2.AuthorizationCodeInstalledApp;
  import com.google.api.client.extensions.jetty.auth.oauth2.LocalServerReceiver;
  import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow;
  import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
--- 1,8 ----
  package fr.inria.spirals.repairnator.serializer.gsheet;
  
  import com.google.api.client.auth.oauth2.Credential;
! import com.google.api.client.extensions.java6.auth.
!     oauth2.AuthorizationCodeInstalledApp;
  import com.google.api.client.extensions.jetty.auth.oauth2.LocalServerReceiver;
  import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow;
  import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
