*** ../corpora/repairnator-corpus/data/50/RepositoryModelResolver.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/164/RepositoryModelResolver.java	2019-10-03 10:36:33.000000000 -0400
***************
*** 109,115 ****
  
      @Override
      public ModelSource resolveModel(Parent parent) throws UnresolvableModelException {
!         return resolveModel(parent.getGroupId(), parent.getArtifactId(), parent.getVersion());
      }
  
      @Override
--- 109,116 ----
  
      @Override
      public ModelSource resolveModel(Parent parent) throws UnresolvableModelException {
!         return resolveModel(parent
!                 .getGroupId(), parent.getArtifactId(), parent.getVersion());
      }
  
      @Override
