--- ../corpora/repairnator-corpus/data/38/CSVSerializerUtils.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/185/CSVSerializerUtils.java	2019-10-03 10:56:14.000000000 -0400
@@ -15,7 +15,8 @@
     private static final Logger logger = LoggerFactory.getLogger(CSVSerializerUtils.class);
 
     public static final char SEPARATOR = ',';
-    private static final String FILENAME = "all_data_builds.csv";
+    private static final String FILENAME
+        = "all_data_builds.csv";
 
     public static BufferedWriter openFile(String outputPath, String firstLine) {
         File outputFile = new File(outputPath);
