--- ../corpora/repairnator-corpus/data/45/MavenOutputHandler.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/19/MavenOutputHandler.java	2019-10-03 10:56:04.000000000 -0400
@@ -30,7 +30,8 @@
 
     private void initFileWriter() {
         String filename = "repairnator.maven." + name.toLowerCase() + ".log";
-        String filePath = inspector.getRepoLocalPath() + "/" + filename;
+        String filePath = inspector
+            .getRepoLocalPath() + "/" + filename;
 
         try {
             this.fileWriter = new FileWriter(filePath);
