*** ../corpora/repairnator-corpus/data/36/SerializerUtils.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/331/SerializerUtils.java	2019-10-03 10:36:36.000000000 -0400
***************
*** 1,4 ****
! package fr.inria.spirals.repairnator.serializer;
  
  import org.apache.commons.lang.time.DurationFormatUtils;
  
--- 1,5 ----
! package fr.inria.spirals.repairnator
!         .serializer;
  
  import org.apache.commons.lang.time.DurationFormatUtils;
  
