--- ../corpora/repairnator-corpus/data/30/GoogleSpreadSheetNopolSerializer.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/143/GoogleSpreadSheetNopolSerializer.java	2019-10-03 10:56:12.000000000 -0400
@@ -103,7 +103,8 @@
                     valueRange.setValues(dataRow);
 
                     try {
-                        AppendValuesResponse response = this.sheets.spreadsheets().values()
+                        AppendValuesResponse response = this.sheets
+                            .spreadsheets().values()
                                 .append(GoogleSpreadSheetFactory.getSpreadsheetID(), RANGE, valueRange)
                                 .setInsertDataOption("INSERT_ROWS").setValueInputOption("USER_ENTERED").execute();
                         if (response != null && response.getUpdates().getUpdatedCells() > 0) {
