--- ../corpora/repairnator-corpus/data/11/MavenLogParser.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/364/MavenLogParser.java	2019-10-03 10:56:24.000000000 -0400
@@ -51,7 +51,8 @@
         Pattern mvnTestHeadPattern = Pattern.compile(MVN_TESTS_PATTERN);
 
         for (String s : outOfFold.getContent()) {
-            Matcher mvnLineMatcher = mvnLinePattern.matcher(s);
+            Matcher
+                    mvnLineMatcher = mvnLinePattern.matcher(s);
 
             if (!beginTestHead && !inTestHead && mvnLineMatcher.matches()) {
                 beginTestHead = true;
