*** ../corpora/repairnator-corpus/data/41/JsonSerializer.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/304/JsonSerializer.java	2019-10-03 10:36:36.000000000 -0400
***************
*** 221,227 ****
          result.addProperty("nbTests", testInformation.getNbTotalTests());
          result.addProperty("nbSkippingTests", testInformation.getNbSkippingTests());
          result.add("errors", serialize(inspector.getStepErrors()));
!         this.inspectors.get(inspector.getState().name()).add(result);
      }
  
      private void outputHasBeenPatchedInspector(ProjectInspector inspector) {
--- 221,228 ----
          result.addProperty("nbTests", testInformation.getNbTotalTests());
          result.addProperty("nbSkippingTests", testInformation.getNbSkippingTests());
          result.add("errors", serialize(inspector.getStepErrors()));
!         this.inspectors.get(inspector.getState().
!         name()).add(result);
      }
  
      private void outputHasBeenPatchedInspector(ProjectInspector inspector) {
