*** ../corpora/repairnator-corpus/data/3/Repository.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/333/Repository.java	2019-10-03 10:36:37.000000000 -0400
***************
*** 1,6 ****
  package fr.inria.spirals.jtravis.entities;
  
! import fr.inria.spirals.jtravis.helpers.BuildHelper;
  import fr.inria.spirals.jtravis.pojos.RepositoryPojo;
  
  /**
--- 1,7 ----
  package fr.inria.spirals.jtravis.entities;
  
! import fr.inria.spirals.jtravis.helpers.
! BuildHelper;
  import fr.inria.spirals.jtravis.pojos.RepositoryPojo;
  
  /**
