*** ../corpora/repairnator-corpus/data/10/PRInformation.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/187/PRInformation.java	2019-10-03 10:36:33.000000000 -0400
***************
*** 43,49 ****
          PRInformation that = (PRInformation) o;
  
          if (head != null ? !head.equals(that.head) : that.head != null) return false;
!         if (base != null ? !base.equals(that.base) : that.base != null) return false;
          return otherRepo != null ? otherRepo.equals(that.otherRepo) : that.otherRepo == null;
      }
  
--- 43,50 ----
          PRInformation that = (PRInformation) o;
  
          if (head != null ? !head.equals(that.head) : that.head != null) return false;
!         if (base != null ? !base
!                 .equals(that.base) : that.base != null) return false;
          return otherRepo != null ? otherRepo.equals(that.otherRepo) : that.otherRepo == null;
      }
  
