--- ../corpora/repairnator-corpus/data/58/ScannedBuildStatus.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/298/ScannedBuildStatus.java	2019-10-03 10:56:21.000000000 -0400
@@ -3,6 +3,7 @@
 /**
  * Created by fernanda on 24/02/17.
  */
-public enum ScannedBuildStatus {
+public enum ScannedBuildStatus
+{
     ONLY_FAIL, PASSING_AND_FAIL, PASSING_AND_PASSING, BUILD_WITH_TEST_CHANGES
 }
