*** ../corpora/repairnator-corpus/data/22/BuildHelperTest.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/399/BuildHelperTest.java	2019-10-03 10:36:38.000000000 -0400
***************
*** 35,40 ****
--- 35,41 ----
          commit.setSha("d283ce5727f47c854470e64ac25144de5d8e6c05");
          commit.setMessage("test: add test for method parameter templating (#1064)");
          commit.setCompareUrl("https://github.com/INRIA/spoon/compare/3c5ab0fe7a89...d283ce5727f4");
+ 
          commit.setBranch("master");
          commit.setAuthorName("Martin Monperrus");
          commit.setAuthorEmail("monperrus@users.noreply.github.com");
***************
*** 154,158 ****
          assertTrue(obtainedBuild != null);
          assertEquals(expectedBuildId, obtainedBuild.getId());
      }
!     
  }
--- 155,159 ----
          assertTrue(obtainedBuild != null);
          assertEquals(expectedBuildId, obtainedBuild.getId());
      }
! 
  }
