--- ../corpora/repairnator-corpus/data/59/ComputeClasspath.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/352/ComputeClasspath.java	2019-10-03 10:56:23.000000000 -0400
@@ -53,7 +53,8 @@
         File defaultClassDir = new File(incriminatedModule + File.separator + DEFAULT_CLASSES_DIR + File.separator);
         this.addFileToClassPath(defaultClassDir);
 
-        File defaultTestClassDir = new File(
+        File
+        defaultTestClassDir = new File(
                 incriminatedModule + File.separator + DEFAULT_TEST_CLASSES_DIR + File.separator);
         this.addFileToClassPath(defaultTestClassDir);
 
