--- ../corpora/repairnator-corpus/data/12/JavaLogParser.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/207/JavaLogParser.java	2019-10-03 10:56:15.000000000 -0400
@@ -21,6 +21,7 @@
         result.setErrored(this.erroredTests);
         result.setRunning(this.runningTests);
         result.setSkipping(this.skippingTests);
+
         result.setPassing(this.passingTests);
         result.setFailing(this.failingTests);
 
