--- ../corpora/repairnator-corpus/data/70/GatherTestInformation.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/400/GatherTestInformation.java	2019-10-03 10:56:26.000000000 -0400
@@ -14,7 +14,8 @@
 import java.util.Locale;
 import java.util.Set;
 
-import org.apache.commons.lang3.StringUtils;
+import org.apache
+    .commons.lang3.StringUtils;
 import org.apache.maven.plugins.surefire.report.ReportTestCase;
 import org.apache.maven.plugins.surefire.report.ReportTestSuite;
 import org.apache.maven.plugins.surefire.report.SurefireReportParser;
