--- ../corpora/repairnator-corpus/data/52/NopolStatus.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/170/NopolStatus.java	2019-10-03 10:56:13.000000000 -0400
@@ -4,5 +4,6 @@
  * Created by urli on 16/02/2017.
  */
 public enum NopolStatus {
-    NOTLAUNCHED, RUNNING, NOPATCH, TIMEOUT, PATCH, EXCEPTION
+    NOTLAUNCHED, RUNNING,
+            NOPATCH, TIMEOUT, PATCH, EXCEPTION
 }
