--- ../corpora/repairnator-corpus/data/66/CloneRepository.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/437/CloneRepository.java	2019-10-03 10:56:28.000000000 -0400
@@ -220,7 +220,8 @@
             Git git = Git.cloneRepository().setURI(repoRemotePath).setDirectory(new File(repoLocalPath)).call();
 
             this.writeProperty("workspace", this.inspector.getWorkspace());
-            this.writeProperty("buildid", this.build.getId() + "");
+            this.
+            writeProperty("buildid", this.build.getId() + "");
             this.writeProperty("repo", this.build.getRepository().getSlug());
 
             if (this.build.isPullRequest()) {
