--- ../corpora/repairnator-corpus/data/66/CloneRepository.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/405/CloneRepository.java	2019-10-03 10:56:26.000000000 -0400
@@ -57,7 +57,8 @@
                 + " Reset hour: " + dateFormat.format(rateLimit.reset));
     }
 
-    private String getLastKnowParent(GitHub gh, GHRepository ghRepo, Git git, String oldCommitSha) throws IOException {
+    private String getLastKnowParent(GitHub gh, GHRepository ghRepo, Git git, String oldCommitSha) throws IOException
+    {
         this.showGitHubRateInformation(gh);
         GHCommit commit = ghRepo.getCommit(oldCommitSha); // get the deleted
                                                           // commit from GH
