--- ../corpora/repairnator-corpus/data/47/MavenFilterTestOutputHandler.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/9/MavenFilterTestOutputHandler.java	2019-10-03 10:56:03.000000000 -0400
@@ -77,7 +77,8 @@
             this.erroringTests += Integer.parseInt(matcher.group(3));
             this.skippingTests += Integer.parseInt(matcher.group(4));
 
-            if (matcher.group(5) != null) {
+            if (matcher.group(5) != null
+        ) {
                 this.totalElapsedTime += Float.parseFloat(matcher.group(6));
 
                 if (failing > 0) {
