*** ../corpora/repairnator-corpus/data/44/ProjectInspector4Bears.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/204/ProjectInspector4Bears.java	2019-10-03 10:36:34.000000000 -0400
***************
*** 55,61 ****
  
                  cloneRepo.setNextStep(buildRepo).setNextStep(testProject).setNextStep(this.testInformations)
                          .setNextStep(checkoutPreviousBuild).setNextStep(buildRepoForPreviousBuild)
!                         .setNextStep(testProjectForPreviousBuild).setNextStep(gatherTestInformation)
                          .setNextStep(checkoutSourceCodeForPreviousBuild).setNextStep(buildRepoForPreviousBuild2)
                          .setNextStep(testProjectForPreviousBuild2).setNextStep(gatherTestInformation2);
  
--- 55,62 ----
  
                  cloneRepo.setNextStep(buildRepo).setNextStep(testProject).setNextStep(this.testInformations)
                          .setNextStep(checkoutPreviousBuild).setNextStep(buildRepoForPreviousBuild)
!                         .setNextStep(testProjectForPreviousBuild).
!                         setNextStep(gatherTestInformation)
                          .setNextStep(checkoutSourceCodeForPreviousBuild).setNextStep(buildRepoForPreviousBuild2)
                          .setNextStep(testProjectForPreviousBuild2).setNextStep(gatherTestInformation2);
  
