--- ../corpora/repairnator-corpus/data/55/ComparatorFailureLocation.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/219/ComparatorFailureLocation.java	2019-10-03 10:56:15.000000000 -0400
@@ -8,7 +8,8 @@
 public class ComparatorFailureLocation implements Comparator<FailureLocation> {
     @Override
     public int compare(FailureLocation o1, FailureLocation o2) {
-        if (o1.equals(o2)) {
+        if (o1.equals(o2
+    )) {
             return 0;
         } else {
             int diffFail = o2.getNbFailures() - o1.getNbFailures();
