--- ../corpora/repairnator-corpus/data/56/FailureLocation.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/356/FailureLocation.java	2019-10-03 10:56:23.000000000 -0400
@@ -68,6 +68,7 @@
 
     @Override
     public String toString() {
-        return "Test case: " + className + " (Fail: " + nbFailures + "; Error: " + nbErrors + ")";
+        return "Test case: " +
+            className + " (Fail: " + nbFailures + "; Error: " + nbErrors + ")";
     }
 }
