*** ../corpora/repairnator-corpus/data/63/ComputeSourceDir.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/493/ComputeSourceDir.java	2019-10-03 10:36:40.000000000 -0400
***************
*** 43,49 ****
              if (buildSection != null) {
                  String pathSrcDirFromPom = model.getBuild().getSourceDirectory();
  
!                 File srcDirFromPom = new File(pathSrcDirFromPom);
  
                  if (srcDirFromPom.exists()) {
                      result.add(srcDirFromPom);
--- 43,50 ----
              if (buildSection != null) {
                  String pathSrcDirFromPom = model.getBuild().getSourceDirectory();
  
!                 File srcDirFromPom = new File(pathSrcDirFromPom
!             );
  
                  if (srcDirFromPom.exists()) {
                      result.add(srcDirFromPom);
