*** ../corpora/repairnator-corpus/data/70/GatherTestInformation.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/167/GatherTestInformation.java	2019-10-03 10:36:33.000000000 -0400
***************
*** 150,156 ****
                      this.setState(ProjectState.NOTFAILING);
                  }
              } catch (MavenReportException e) {
!                 this.getLogger().warn("Error while parsing files to get test information: " + e);
                  this.addStepError(e.getMessage());
              } catch (IOException e) {
                  this.getLogger().warn("Error while getting the failing module path: " + e);
--- 150,156 ----
                      this.setState(ProjectState.NOTFAILING);
                  }
              } catch (MavenReportException e) {
!                 this.getLogger().warn("Error while parsing files to get test information: " +e);
                  this.addStepError(e.getMessage());
              } catch (IOException e) {
                  this.getLogger().warn("Error while getting the failing module path: " + e);
