--- ../corpora/repairnator-corpus/data/15/BuildHelper.java	2019-01-09 09:23:14.000000000 -0500
+++ ./tmp/ffe9c0e978da4930a8b4f9cdf1fb1d6e/179/BuildHelper.java	2019-10-03 10:56:14.000000000 -0400
@@ -53,7 +53,8 @@
         String resourceUrl = getInstance().getEndpoint()+BUILD_ENDPOINT+id;
 
         try {
-            String response = getInstance().get(resourceUrl);
+            String response = getInstance().get(
+                resourceUrl);
             JsonParser parser = new JsonParser();
             JsonObject allAnswer = parser.parse(response).getAsJsonObject();
 
