*** ../corpora/repairnator-corpus/data/56/FailureLocation.java	2019-01-09 09:23:14.000000000 -0500
--- ./tmp/085a35be431e49c792cd6b10e9e8a9f6/486/FailureLocation.java	2019-10-03 10:36:40.000000000 -0400
***************
*** 60,66 ****
  
          return className != null ? className.equals(that.className) : that.className == null;
      }
- 
      @Override
      public int hashCode() {
          return className != null ? className.hashCode() : 0;
--- 60,65 ----
